shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs

```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
+        data_center: &str,
+    ) -> Result<HashMap<String, KeyspaceMetadata>> {
+        let (tx, rx) = oneshot::channel();
+
+        connection.send(Message::from_frame(
+            Frame::Cassandra(CassandraFrame{
+                version: Version::V4,
+                stream_id: 0,
+                tracing: Tracing::Request(false),
+                warnings: vec![],
+                operation: CassandraOperation::Query{
+                    query: Box::new(
+                        parse_statement_single(
+
+                            "SELECT keyspace_name, toJson(replication) AS replication FROM system_schema.keyspaces",
+                        )
+                    ),
+
+                    params: Box::new(QueryParams::default()),
+                }
+            })),
+            tx
+        )?;
+
+        let response = rx.await?.response?;
+        into_keyspaces(response, data_center)
+    }
+
+    fn into_keyspaces(
+        mut response: Message,
+        data_center: &str,
+    ) -> Result<HashMap<String, KeyspaceMetadata>> {
+        if let Some(Frame::Cassandra(frame)) = response.frame() {
+            match &mut frame.operation {
+                CassandraOperation::Result(CassandraResult::Rows { rows, .. }) => rows
+                    .iter_mut()
+                    .map(|row| build_keyspace(row, data_center))
+                    .try_collect(),
+                operation => Err(anyhow!(
+                    "keyspace query returned unexpected cassandra operation: {:?}",
+                    operation
+                )),
+            }
+        } else {
+            Err(anyhow!("Failed to parse keyspace query response"))
+        }
+    }
+
+    fn build_keyspace(
+        row: &mut Vec<MessageValue>,
+        data_center: &str,
+    ) -> Result<(String, KeyspaceMetadata)> {
+        let metadata = if let Some(MessageValue::Varchar(string)) = row.pop() {
+            let replication: JsonValue = serde_json::from_str(&string).map_err(|error| {
+                anyhow!(format!(
+                    "Error parsing replication. Error: {} Replication: {}",
+                    error, string
+                ))
+            })?;
+
+            let replication_strategy = match replication {
+                JsonValue::Object(properties) => build_replication_strategy(properties)?,
+                _ => {
+                    return Err(anyhow!(format!(
+                        "Error parsing replication strategy: {}",
+                        replication
+                    )))
+                }
+            };
+
+            let replication_factor = match replication_strategy {
+                ReplicationStrategy::SimpleStrategy { replication_factor } => replication_factor,
+                ReplicationStrategy::NetworkTopologyStrategy {
+                    datacenter_replication_factor,
+                } => *datacenter_replication_factor.get(data_center).unwrap_or(&0),
+                _ => 0,
```

Comment:
Should this be 1 instead of 0?
I imagine 1 means the data is stored in one location?

Reply:
The answer is that because the replication factor is per dc, we can have a replication factor of 0 in one dc while having a non-zero replication factor in other dc's, so 0 does make semantic sense.

I''m still a little concerned about this line for other reasons but that is addressed in https://github.com/shotover/shotover-proxy/pull/895/files#r1012544156